dragonfly/actions/action_text.py

```diff
@@ -3,109 +3,178 @@
+       spacing and capitalization.  This is done by first mimicking
+       a word recognition and then analyzing its spacing and
+       capitalization and applying the same formatting to the text.
     """
 
     _pause_default = 0.02
     _keyboard = Keyboard()
     _specials = {
-                 "\n":   typeables["enter"],
-                 "\t":   typeables["tab"],
+                 "\n": typeables["enter"],
+                 "\t": typeables["tab"],
                 }
 
     def __init__(self, spec=None, static=False, pause=_pause_default,
-                 autofmt=False):
+                 autofmt=False, use_hardware=False):
         self._pause = pause
         self._autofmt = autofmt
+        self._use_hardware = use_hardware
         DynStrActionBase.__init__(self, spec=spec, static=static)
 
     def _parse_spec(self, spec):
-        """ Convert the given *spec* to keyboard events. """
+        """Convert the given *spec* to keyboard events."""
+        from struct import unpack
         events = []
-        for character in spec:
-            if character in self._specials:
-                typeable = self._specials[character]
-            else:
-                try:
-                    typeable = Keyboard.get_typeable(character)
-                except ValueError as e:
-                    raise ActionError("Keyboard interface cannot type this"
-                                      " character: %r (in %r)"
-                                      % (character, spec))
-            events.extend(typeable.events(self._pause))
+        if self._use_hardware or require_hardware_emulation():
+            for character in spec:
+                if character in self._specials:
+                    typeable = self._specials[character]
+                    events.extend(typeable.events(self._pause))
+                else:
+                    try:
+                        typeable = Keyboard.get_typeable(character)
+                        events.extend(typeable.events(self._pause))
+                    except ValueError:
+                        raise ActionError("Keyboard interface cannot type this"
+                                          " character: %r (in %r)"
+                                          % (character, spec))
+        else:
+            for character in text_type(spec):
+                if character in self._specials:
+                    typeable = self._specials[character]
+                    events.extend(typeable.events(self._pause))
+                else:
+                    byte_stream = character.encode("utf-16-le")
+                    for short in unpack("<" + str(len(byte_stream) // 2) + "H",
+                                        byte_stream):
+                        try:
+                            typeable = Keyboard.get_typeable(short,
+                                                             is_text=True)
+                            events.extend(typeable.events(self._pause * 0.5))
```

Comment:
I empirically pulled that figure entirely out of my neither region. Considering the way Windows implements Unicode keypresses, it could realistically just be zero. More on that later.